deepchem/models/tensorgraph/tests/test_tensor_graph.py

```diff
@@ -1,50 +1,254 @@
+import tensorflow as tf
+import os
+
+from deepchem.data import NumpyDataset
+from deepchem.data.datasets import Databag
+from deepchem.models.tensorgraph.layers import Input, Dense, LossLayer, Flatten, ReduceSquareDifference
+from deepchem.models.tensorgraph.layers import Layer, Input, Reshape, Flatten, Feature, Conv2d, MaxPool, Label
+from deepchem.models.tensorgraph.layers import Dense, SoftMaxCrossEntropy, ReduceMean, SoftMax
 from deepchem.models.tensorgraph.tensor_graph import TensorGraph
-from nose.tools import nottest
+from nose.tools import nottest, assert_true, assert_equals
+from sklearn.metrics import roc_curve, auc
+import numpy as np
 
 
 class TestTensorGraph(unittest.TestCase):
   """
   Test that graph topologies work correctly.
   """
 
-  @nottest
-  def test_graph_save(self):
-    n_samples = 10
-    n_features = 11
-    n_tasks = 1
-    batch_size = 10
-    X = np.random.rand(batch_size, n_samples, n_features)
-    y = np.ones(shape=(n_samples, n_tasks))
-    ids = np.arange(n_samples)
-
-    dataset = dc.data.NumpyDataset(X, y, None, ids)
-    g = TensorGraph(model_dir='/tmp/tmpss5_ki5_')
-
-    inLayer = Input(shape=(None, n_samples, n_features))
-    g.add_feature(inLayer)
-
-    flatten = Flatten()
-    g.add_layer(flatten, parents=[inLayer])
-
-    dense = Dense(out_channels=1)
-    g.add_layer(dense, parents=[flatten])
-    g.add_output(dense)
-
-    label_out = Input(shape=(None, 1))
-    g.add_label(label_out)
-
-    loss = LossLayer()
-    g.add_layer(loss, parents=[dense, label_out])
-    g.set_loss(loss)
-
-    g.fit(dataset, nb_epoch=100)
-    g.save()
-    g1 = TensorGraph.load_from_dir('/tmp/tmpss5_ki5_')
-    print(g1)
-    print(g1.predict_on_batch(X))
+  def test_single_task_classifier(self):
+    n_data_points = 200
+    n_features = 10
+    X = np.random.rand(n_data_points, n_features)
+    y = [[0, 1] for x in range(n_data_points)]
+    dataset = NumpyDataset(X, y)
+    features = Feature(shape=(None, n_features))
+    dense = Dense(out_channels=2, in_layers=[features])
+    output = SoftMax(in_layers=[dense])
+    label = Label(shape=(None, 2))
+    smce = SoftMaxCrossEntropy(in_layers=[label, dense])
+    loss = ReduceMean(in_layers=[smce])
+    tg = dc.models.TensorGraph(learning_rate=0.1)
+    tg.add_output(output)
+    tg.set_loss(loss)
+    tg.fit(dataset, nb_epoch=100)
+    prediction = np.squeeze(tg.predict_proba_on_batch(X))
+    assert_true(np.all(np.isclose(prediction, y, atol=0.05)))
+
+  def test_multi_task_classifier(self):
+    n_data_points = 20
+    n_features = 10
+
+    X = np.random.rand(n_data_points, n_features)
+    y1 = np.array([[0, 1] for x in range(n_data_points)])
+    y2 = np.array([[1, 0] for x in range(n_data_points)])
+    X = NumpyDataset(X)
+    ys = [NumpyDataset(y1), NumpyDataset(y2)]
+
+    databag = Databag()
+
+    features = Feature(shape=(None, n_features))
+    databag.add_dataset(features, X)
+
+    outputs = []
+    entropies = []
+    for i in range(2):
```

Comment:
Our current convention of having a separate placeholder for each label is a little awkward (as @lilleswing has mentioned offline before :-)). In a follow-up PR, let's make a `MultiOutput` layer that makes it easy to build models with many output heads. This should take in only a single `Label` of shape `(None, n_tasks)` or `(None, n_tasks, n_classes)`